### DISCUSSION TITLE
Running OceanParcels with surface velocity data from Copernicus.

### DISCUSSION THREAD

**Discussion by claudioiturra** [Parcels-code/Parcels#1530] [General]
Hello, I'm new to Ocean-Parcels, sorry! I'm currently attempting to run a Lagrangian particle distribution using velocities from Copernicus (.nc file). Here is my [code](https://github.com/claudioiturra/desktop-tutorial/blob/main/Untitled.ipynb). The .nc file contains the variables uo, vo, longitude, latitude, and time (hours since 1950-01-01, 2137hours). As described in the code, I got the error in the runtime and field sampe out of bond, I will appreciate any suggestion, is actually de code right?. 

FieldOutOfBoundError                      Traceback (most recent call last)
Cell In[7], line 7
      1 output_file = pset.ParticleFile(
      2     name="GCParticles.zarr",  # the name of the output file
      3     outputdt=3600,  # the time period between consecutive out output steps
      4     chunks=(1, 10),  # the chunking of the output file (number of particles, timesteps)
      5 )
----> 7 pset.execute(
      8     kernels,  # the kernel (which defines how particles move)
      9     runtime=86400 * 24,  # the total length of the run in seconds
     10     dt=300,  # the timestep of the kernel in seconds
     11     output_file=output_file,
     12 )

File [/opt/anaconda3/envs/parcels/lib/python3.12/site-packages/parcels/particleset.py:951](http://localhost:8888/opt/anaconda3/envs/parcels/lib/python3.12/site-packages/parcels/particleset.py#line=950), in ParticleSet.execute(self, pyfunc, pyfunc_inter, endtime, runtime, dt, output_file, verbose_progress, postIterationCallbacks, callbackdt)
    949 # If we don't perform interaction, only execute the normal kernel efficiently.
    950 if self.interaction_kernel is None:
--> 951     res = self.kernel.execute(self, endtime=next_time, dt=dt)
    952     if res == StatusCode.StopAllExecution:
    953         return StatusCode.StopAllExecution

File [/opt/anaconda3/envs/parcels/lib/python3.12/site-packages/parcels/kernel.py:585](http://localhost:8888/opt/anaconda3/envs/parcels/lib/python3.12/site-packages/parcels/kernel.py#line=584), in Kernel.execute(self, pset, endtime, dt)
    583     raise TimeExtrapolationError(p.time)
    584 elif p.state == StatusCode.ErrorOutOfBounds:
--> 585     raise FieldOutOfBoundError(p.lon, p.lat, p.depth)
    586 elif p.state == StatusCode.ErrorThroughSurface:
    587     raise FieldOutOfBoundSurfaceError(p.lon, p.lat, p.depth)

FieldOutOfBoundError: Field sampled out-of-bound, at (-73.2887954711914, -37.59098434448242, 0.0)

Variables:
    uo       
           Size:       21x28x1x8748
           Dimensions: longitude,latitude,depth,time
           Datatype:   single
           Attributes:

**Comment by erikvansebille** [comment]
Hi @claudioiturra, thanks for using Parcels!

Your error is a `FieldOutOfBoundError`, which means that you try to access a location in the ocean (in your case 73W, 37S) that is not part of the domain. See [this part of the tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes) on how to work with this and other `ErrorCodes`

**Reply by claudioiturra** [reply]
Thanks @erikvansebille for the response. Finally, it was a problem with the setup. After multiple attempts, I could run the OceanParcels. However, I have encountered a new problem. After running OceanParcels (as shown in the figure on the left), it appears that the trajectory suddenly veers southward before returning northward. However, when I extract the location and run the trajectory in MATLAB (figure right), it does not exhibit a similar pattern to the OceanParcels result.
![Screenshot from 2024-04-04 12-00-50](https://github.com/OceanParcels/parcels/assets/61473481/a127fdde-a07b-449e-8661-06112630f925)

Here's my code https://github.com/claudioiturra/desktop-tutorial/blob/main/makeFIX.ipynb
 I will super appreciate any help on how to fix that error.

**Reply by michaeldenes** [reply]
Hi @claudioiturra, I've just had a look at your linked notebook. I think the issue is in your `NorthVel` kernel, what exactly is the purpose of this kernel? Parcels works by converting m/s values into deg/s (scaled with latitude accordingly).

In your case, between times 10x86400, and 10.2x86400 (that is, t in [864000, 881280], or a total time of 17280 seconds), you're moving your particles south by 0.001x3600 = 0.36 degrees per timestep). This means, between t in [864000, 881280], there are 4.8 timesteps (so you'll only experience 4 timesteps in total) you're shifting your particle 1.44 degrees south using your `NorthVel` kernel. So your plot on the left looks to be doing exactly what your kernel is coded for.

If you're looking to trace a water parcel, I would remove the `NorthVel` kernel.

**Reply by claudioiturra** [reply]
Hi @michaeldenes , yes.. thank you so much! Actually I was learning about kernels during the night, now everything works great. Such a great tool, Im currently learning a lot using the tutorial. Thanks again.

**Reply by claudioiturra** [reply]
Hello again! Well... thank you soooooo much for all the help @michaeldenes @erikvansebille ! Now I have the last one! I'm not that good with Python. I will appreciate any help on how to plot each trajectory with different colors and add a point at a specific location. 

ds = xr.open_zarr("JAN2023GULFARAUCO.zarr")
ds.traj.plot(margin=2)
#x = [-73.1224] 
#y = [-36.4075]
#plt.show() Did not work
plt.plot(x, y, marker="o", markersize=20, markeredgecolor="red", markerfacecolor="green")

Thank for your time guys...

![Screenshot from 2024-04-05 17-27-10](https://github.com/OceanParcels/parcels/assets/61473481/e7075a6b-6e97-47bb-921a-e3e7cfca74a2)

**Reply by michaeldenes** [reply]
Hi @claudioiturra, no problem, we're glad you're enjoying using the tool. There's quite a few tutorials on our webpage (see, https://docs.oceanparcels.org/en/latest/documentation.html ), but the specific one that might help you is https://docs.oceanparcels.org/en/latest/examples/tutorial_output.html .

Specifically, if `ds` is your trajectory data, then the following should work nicely:
`plt.plot(ds["lon"].T, ds["lat"].T)`

To plot a point at a specific location (e.g., the start points of the trajectories), you can do a simple scatter plot similar to above. But all of this is can be determined from the tutorial.

Cheers!

**Reply by claudioiturra** [reply]
Thanks @michaeldenes , finally I solve that with a for. Looks great!
<img width="606" alt="Screenshot 2024-04-07 at 10 41 35 AM" src="https://github.com/OceanParcels/parcels/assets/61473481/4edc0a52-cda7-4ace-b0dd-45c1845ddaf4">

**Reply by michaeldenes** [reply]
@claudioiturra Perfect! Happy to have helped :)
